WIP on hash group joiner

yuzefovich opened the pull request:
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101181)
<!-- Reviewable:end -->

yuzefovich committed:
WIP on optimized hash group joiner

yuzefovich committed:
WIP on extracting JoinProbeState

yuzefovich committed:
colexecjoin: extract HashJoinerSpec into colexecargs

This commit extracts `HashJoinerSpec` struct from `colexecjoin` into
`colexecargs` as well as exports a couple of its fields. This is needed
in order to reuse this struct by the hash group joiner that lives in
`colexec` package without having to make `colexec` depend on
`colexecjoin`.

Release note: None

yuzefovich committed:
WIP on tests

yuzefovich committed:
WIP on FindBuckets incorrect usage

yuzefovich committed:
Revert "WIP on tests"

This reverts commit c29bac3dd1d763868a1f1cfec9c6727b456d2893.
